### PR TITLE
[Android] ES6 & API 19-23 compatibility changes

### PR DESCRIPTION
**Pull request by jahorton**
This will need a 🍒-pick to the `master` branch.

Addresses #2323.

----

Everything's in a pretty "rough-draft" state, as I'm still working on a baseline compat fix that will also handle APIs below 23.

Among other things, referencing the JS `console` from the WebWorker was also generating errors that could crash KMW.

### CODE REVIEW

File: web/source/dom/utils.ts [hunk 1]
```diff
@@ -26,8 +26,8 @@ namespace com.keyman.dom {
         }
 
         // On mobile devices, the OSK uses 'fixed' - this requires some extra offset work to handle.
-        if(Lobj.style.position == 'fixed') {
-          let Ldoc = Lobj.ownerDocument;
+        let Ldoc = Lobj.ownerDocument;
+        if(Lobj.style.position == 'fixed' && Ldoc && Ldoc.scrollingElement) {
```
Comment (mcdurdin) [inline web/source/dom/utils.ts]: I take it this is for compatibility with earlier versions of DOM?

Comment (jahorton) [reply web/source/dom/utils.ts]: Yes - an error about how `scrollingElement` was undefined turned up in the console logs on some of the older APIs.

**Comment by jahorton** [comment]
The remaining error causing instability looks to be trickier to diagnose - we're getting an error when loading the lexical model script on API 19 and 21.

All we get is the generic "Script error" log message, and there's nothing immediately obvious in the script that would break.  More investigation required.

**Comment by jahorton** [comment]
Still figuring out exactly where the issue is.  Some interesting notes:

- The LMLayer itself initializes fine in API 21 (and probably 19).
- KMW then requests that the LMLayer load a model; it's the `importScripts` call that is failing.
- The expected scope (i.e, `LMLayerWorker.loadModel`) exists; the issue is occurring at some point within the model initialization process.
- The issue is _not_ due to `Array.prototype.map`, which is used within `models.TrieModel`.  (Unless it's only missing within WebWorkers; it exists just fine in the standard DOM JS engine)
- The `let... of` construct is also fine.

I'm pretty sure we're using _something_ within the `TrieModel` class that's not supported in the JS engine for these old APIs, but I haven't yet found exactly what that _something_ is.  We only receive the generic log message "Uncaught error:  Script error" out of the `importScripts` call due to technically cross-origin loading.  (blob: vs file:)

**Comment by jahorton** [comment]
After way too much investigation time and jerry-rigging test setups, I've finally found an answer to the last problem - the ES5 engine doesn't like duplicate property names on objects, and there's a minor bug generating them within the lexical model compiler.

In particular, two SMP characters have the same leading surrogate:
- 🇸
- 🇺

It appears that when compiling SMP strings for Trie models, a node is being generated for the leading surrogate, followed by a node for the completed surrogate pair.  Unfortunately, a unique copy of the leading surrogate is generated for each surrogate pair, allowing duplicate object properties to exist if two SMP characters (like the two above) have the same leading surrogate.

**Comment by jahorton** [comment]
The current changes in this PR's branch, when merged with the current `12.0-stable` and `web-2379-cherry-macos-firefox-versions`, passes unit tests - a pretty good sign.  (Both need to be merged in for CI testing success.)

File: common/predictive-text/worker/index.ts [hunk 2]
```diff
@@ -123,7 +131,10 @@ class LMLayerWorker {
     if(im.message == 'load') {
       let data = im as LoadMessage;
       if(data.model == this._currentModelSource) {
-        console.warn("Duplicate model load message detected - squashing!");
+        // Some JS implementations don't allow web workers access to the console.
+        if(console) {
```
Comment (darcywong00) [inline common/predictive-text/worker/index.ts]: This is post-mortem. I got a KMW error `Uncaught ReferenceError: console is not defined`.
Should this be
```suggestion
        if(typeof console !== 'undefined') {
```

Comment (jahorton) [reply common/predictive-text/worker/index.ts]: Yes.  It was supposed to be in there; maybe I accidentally removed it when I had started to splinter the PR in two, then decided against it.

Comment (mcdurdin) [reply common/predictive-text/worker/index.ts]: Good catch @darcywong00 . Will you turn this into a PR?

Comment (darcywong00) [reply common/predictive-text/worker/index.ts]: I've added it to wip #2389 and will park it for now.

**Pull request merged**